ts/src/pro/xt.ts

```diff
@@ -0,0 +1,1079 @@
+        if (marketId !== undefined) {
+            const trade = this.parseTrade (data);
+            const market = this.market (trade['symbol']);
+            const tradeType = market['contract'] ? 'contract' : 'spot';
+            const symbol = market['symbol'];
+            const event = this.safeString (message, 'event');
+            let tradesArray = this.safeValue (this.trades, symbol);
+            if (tradesArray === undefined) {
+                const tradesLimit = this.safeInteger (this.options, 'tradesLimit', 1000);
+                tradesArray = new ArrayCache (tradesLimit);
+                this.trades[symbol] = tradesArray;
+            }
+            tradesArray.append (trade);
+            const messageHash = event + ':' + tradeType;
+            client.resolve (tradesArray, messageHash);
+        }
+        return message;
+    }
+
+    handleOrderBook (client: Client, message) {
+        //
+        // spot
+        //
+        //    {
+        //        "topic": "depth",
+        //        "event": "depth@btc_usdt,20",
+        //        "data": {
+        //            "s": "btc_usdt",        // symbol
+        //            "fi": 1681433733351,
+        //            "i": 1681433733371,
+        //            "a": [                  // asks(sell order)
+        //                [                   // [0]price, [1]quantity
+        //                    "34000",        // price
+        //                    "1.2"           // quantity
+        //                ],
+        //                [
+        //                    "34001",
+        //                    "2.3"
+        //                ]
+        //            ],
+        //            "b": [                   // bids(buy order)
+        //                [
+        //                    "32000",
+        //                    "0.2"
+        //                ],
+        //                [
+        //                    "31000",
+        //                    "0.5"
+        //                ]
+        //            ]
+        //        }
+        //    }
+        //
+        // contract
+        //
+        //    {
+        //        "topic": "depth",
+        //        "event": "depth@btc_usdt,20",
+        //        "data": {
+        //            s: "btc_usdt",
+        //            pu: "548111455664",
+        //            fu: "548111455665",
+        //            u: "548111455667",
+        //            a: [
+        //                [
+        //                    "26841.5",
+        //                    "50210",
+        //                ],
+        //            ],
+        //            b: [
+        //                [
+        //                    "26841",
+        //                    "67075",
+        //                ],
+        //            ],
+        //            t: 1684530667083,
+        //        }
+        //    }
+        //
+        const data = this.safeValue (message, 'data');
```

Comment:
First we're not handling the orderbook synchronization with the rest snapshot (check for instance how Gate does it). Second, the orderbook is missing the symbol and nonce values

```
}
OrderBook {
  bids: Bids(2) [ [ 27810.91, 2.055121 ], [ 27782.2, 0.937568 ] ],
  asks: Asks(7) [
    [ 27816.1, 1.967926 ],
    [ 27819.15, 3.759403 ],
    [ 27858.81, 8.927001 ],
    [ 27915.97, 2.269404 ],
    [ 27916.18, 7.9646 ],
    [ 27932.36, 1.332395 ],
    [ 27932.53, 7.777499 ]
  ],
  timestamp: undefined,
  datetime: undefined,
  nonce: undefined,
  symbol: undefined
}
```

Reply:
Is this the part you're referring to? I don't understand whats happening here. Could you explain a little bit what orderbook synchronization would look like?

```
if (nonce === undefined) {
            let cacheLength = 0;
            if (storedOrderBook !== undefined) {
                cacheLength = storedOrderBook.cache.length;
            }
            const snapshotDelay = this.handleOption ('watchOrderBook', 'snapshotDelay', 10);
            const waitAmount = isSpot ? snapshotDelay : 0;
            if (cacheLength === waitAmount) {
                // max limit is 100
                const subscription = client.subscriptions[messageHash];
                const limit = this.safeInteger (subscription, 'limit');
                this.spawn (this.loadOrderBook, client, messageHash, symbol, limit);
            }
            storedOrderBook.cache.push (delta);
            return;
   ```

Reply:
@samgermain Exactly. Similarly to other exchanges, upon listening to the orderbook stream to get the deltas, we also need to fetch the snapshot through rest, after getting the snapshot we initialize the orderbook itself and start applying the deltas we received while we were waiting. The process should be very similar to the one we do in kucoin, kucounfutures, bitstam, mexc, etc so it should be easy to replicate it here. 


![image](https://github.com/ccxt/ccxt/assets/43336371/f6e4ab57-51de-435b-aae6-9c1730e2d6c7)

Reply:
but we only do it for `watchOrderBook` and don't do it with other methods?

Reply:
> but we only do it for `watchOrderBook` and don't do it with other methods?

@samgermain yes usually that's the only topic that requires it

Reply:
I think I added what you're talking about, but I don't know what these lines mean from the api docs

> 4.Drop any event where i is <= lastUpdateId in the snapshot.
> 5.The first processed event should have fi <= lastUpdateId+1 AND i >= lastUpdateId+1.

What does it mean "Drop any event"

---------

I also don't understand a few parts of the code I added

```
if (cacheLength === snapshotDelay) { // what exactly are we comparing?
       this.spawn (this.loadOrderBook, client, messageHash, symbol); // is this where we get the rest snapshot? Why not just use something like this.fetchOrderBook? 
}
orderbook.cache.push (data); // why are we caching the whole data object? what happens when we do this?
```